File: MATTHEW/leetcode/AIcoding/3.DeepSeek-v3/javascript_solutions/1771-maximize-palindrome-length-from-subsequences.js
var longestPalindrome = function(word1, word2) {
    const s = word1 + word2;
    const n = s.length;
    const m = word1.length;
    let dp = Array.from({ length: n }, () => new Array(n).fill(0));
    let max = 0;

    for (let i = n - 1; i >= 0; i--) {
        dp[i][i] = 1;
        for (let j = i + 1; j < n; j++) {
            if (s[i] === s[j]) {
                dp[i][j] = dp[i + 1][j - 1] + 2;
                if (i < m && j >= m) {
                    max = Math.max(max, dp[i][j]);
                }
            } else {
                dp[i][j] = Math.max(dp[i + 1][j], dp[i][j - 1]);
            }
        }
    }

    return max;
};

File: leetcode/AIcoding/3.DeepSeek-v3/java_solutions/1771-maximize-palindrome-length-from-subsequences.java
class Solution {
    public int longestPalindrome(String word1, String word2) {
        String s = word1 + word2;
        int n = s.length();
        int[][] dp = new int[n][n];
        int res = 0;

        for (int i = n - 1; i >= 0; i--) {
            dp[i][i] = 1;
            for (int j = i + 1; j < n; j++) {
                if (s.charAt(i) == s.charAt(j)) {
                    dp[i][j] = dp[i + 1][j - 1] + 2;
                    if (i < word1.length() && j >= word1.length()) {
                        res = Math.max(res, dp[i][j]);
                    }
                } else {
                    dp[i][j] = Math.max(dp[i + 1][j], dp[i][j - 1]);
                }
            }
        }

        return res;
    }
}